### PR TITLE
Add GDALGetOutputDriversForDatasetName()

### PR DESCRIPTION
**Pull request by rouault**
should help implementing https://github.com/georust/gdal/pull/510

### CODE REVIEW

**Review by lnicola — APPROVED** [review]

File: gcore/gdaldriver.cpp [hunk 1]
```diff
@@ -2810,3 +2810,162 @@ CPLErr GDALDriver::SetMetadataItem(const char *pszName, const char *pszValue,
...
+                    aosDriverNames.AddString(GDALGetDriverShortName(hDriver));
+                }
+            }
+        }
+    }
+
+    // GMT is registered before netCDF for opening reasons, but we want
+    // netCDF to be used by default for output.
+    if (EQUAL(osExt.c_str(), "nc") && aosDriverNames.size() == 2 &&
+        EQUAL(aosDriverNames[0], "GMT") && EQUAL(aosDriverNames[1], "NETCDF"))
```
Comment (lnicola) [inline gcore/gdaldriver.cpp:2924]: Maybe

```suggestion
        EQUAL(aosDriverNames[0], "GMT") && EQUAL(aosDriverNames[1], "netCDF"))
```

I think the short name is `netCDF` and the comparison doesn't seem to be case-insensitive.

Comment (rouault) [reply gcore/gdaldriver.cpp:2924]: > and the comparison doesn't seem to be case-insensitive.

EQUAL() actually does case insensitive comparison, but taking your change.

Comment (Atreyagaurav) [reply gcore/gdaldriver.cpp]: I actually had checked it before and saw that EQUAL macro does case-insensitive in linux/posix compliant ones and case-sensitive one in windows. I was thinking about commenting that but thought it should be fine :D

You can see that if you search for the macro definition. https://github.com/OSGeo/gdal/blob/15751e3b42af9be91e344e7cd4b26c45ed8074ed/frmts/dted/dted_api.h#L56


@lnicola that was also the reason I did, `to_uppercase()` before doing the comparisons (to mirror EQUAL), but you seem to have removed it. Which should still be fine. 

The case for ElasticSearch was changed in some versions, if that happens and the check if not case-insensitive they might fail, but I think it's unlikely.

File: gcore/gdaldriver.cpp [hunk 2]
```diff
@@ -2810,3 +2810,162 @@ CPLErr GDALDriver::SetMetadataItem(const char *pszName, const char *pszValue,
...
+    bool bRet = false;
+    const char *pszDriverExtensions =
+        GDALGetMetadataItem(hDriver, GDAL_DMD_EXTENSIONS, nullptr);
+    if (pszDriverExtensions)
+    {
+        const CPLStringList aosTokens(CSLTokenizeString(pszDriverExtensions));
+        const int nTokens = aosTokens.size();
+        for (int j = 0; j < nTokens; ++j)
+        {
+            if (EQUAL(pszExt, aosTokens[j]))
```
Comment (elpaso) [inline gcore/gdaldriver.cpp:2829]: Is this case sensitive?

File: gcore/gdaldriver.cpp [hunk 3]
```diff
@@ -2810,3 +2810,162 @@ CPLErr GDALDriver::SetMetadataItem(const char *pszName, const char *pszValue,
...
+/*                  GDALGetOutputDriversForDatasetName()                */
+/************************************************************************/
+
+/** Return a list of driver short names that are likely candidates for the
+ * provided output file name.
+ *
+ * @param pszDestDataset Output dataset name (might not exist).
+ * @param nFlagRasterVector GDAL_OF_RASTER, GDAL_OF_VECTOR or
+ *                          binary-or'ed combination of both
+ * @param bSingleMatch Whether a single match is desired. In this mode, if
```
Comment (elpaso) [inline gcore/gdaldriver.cpp:2849]: To be honest I don't understand the naming of this argument and it's doc string. What has the described behavior to do with "single match" ?

File: gcore/gdaldriver.cpp [hunk 4]
```diff
@@ -2810,3 +2810,162 @@ CPLErr GDALDriver::SetMetadataItem(const char *pszName, const char *pszValue,
...
+ * To be freed with CSLDestroy()
+ * @since 3.9
+ */
+char **GDALGetOutputDriversForDatasetName(const char *pszDestDataset,
+                                          int nFlagRasterVector,
+                                          bool bSingleMatch, bool bEmitWarning)
+{
+    CPLStringList aosDriverNames;
+
+    std::string osExt = CPLGetExtension(pszDestDataset);
```
Comment (elpaso) [inline gcore/gdaldriver.cpp:2864]: I have checked CPLGetExtension doc and it doesn't mention case changes. Should it be changed to lowercase for the check down below?

File: gcore/gdaldriver.cpp (hunk 2 above)
Comment (rouault) [reply gcore/gdaldriver.cpp:2829]: No, EQUAL() performs case insensitive comparison

File: gcore/gdaldriver.cpp (hunk 4 above)
Comment (rouault) [reply gcore/gdaldriver.cpp:2864]: > Should it be changed to lowercase for the check down below?

no, cf above comment about EQUAL() being case insensitive

File: gcore/gdaldriver.cpp (hunk 3 above)
Comment (rouault) [reply gcore/gdaldriver.cpp]: Description changed to
```
 * @param bSingleMatch Whether a single match is desired, that is to say the
 *                     returned list will contain at most one item, which will
 *                     be the first driver in the order they are registered to
 *                     match the output dataset name. Note that in this mode, if
 *                     nFlagRasterVector==GDAL_OF_RASTER and pszDestDataset has
 *                     no extension, GTiff will be selected.
 ```

**Review by elpaso — APPROVED** [review]

**Pull request merged**